### PR TITLE
refactor: Use ordered json

### PR DESCRIPTION
**Pull request by Corentin-Allaire**
This update the nlohmann json version to v3.9.1. It then use the new ordered_json feature in the  GeometryHierarchyMapJsonConverter to order the key by order of addition so that they would appear as : 
volume
boundary
layer
approach
sensitive
value

The element in value will still use the normal nlohmann::json and will thus have the keys ordered alphabetically.

### CODE REVIEW

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp [hunk 1]
```diff
@@ -104,16 +104,17 @@ class GeometryHierarchyMapJsonConverter {
 // implementations
 
 template <typename value_t>
-nlohmann::json GeometryHierarchyMapJsonConverter<value_t>::toJson(
+nlohmann::ordered_json GeometryHierarchyMapJsonConverter<value_t>::toJson(
     const Container& container) const {
   // encode header
-  nlohmann::json encoded = nlohmann::json::object();
+  nlohmann::ordered_json encoded = nlohmann::json::object();
+  ;
```
Comment (msmk0) [inline Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:111]:
```suggestion
```
Leftover?

**Review by asalzburger — APPROVED** [review]
Looks good.

**Comment by asalzburger** [comment]
Waiting for the CI to build of course.

**Comment by Corentin-Allaire** [comment]
Once it is in I will update the other PR to take those change into account

File: thirdparty/nlohmann_json/cmake/download_test_data.cmake [hunk 2]
```diff
@@ -0,0 +1,56 @@
+set(JSON_TEST_DATA_URL     https://github.com/nlohmann/json_test_data)
```
Comment (msmk0) [inline thirdparty/nlohmann_json/cmake/download_test_data.cmake:1]: Is this file really needed? I think we are not shipping the tests anyway.

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp [hunk 3]
```diff
@@ -52,7 +52,7 @@ class GeometryHierarchyMapJsonConverter {
   ///
   /// @param container Geometry hierarchy map that should be encoded
   /// @return Encoded Json object
-  nlohmann::json toJson(const Container& container) const;
+  nlohmann::ordered_json toJson(const Container& container) const;
```
Comment (msmk0) [inline Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:55]: Is the use of `ordered_json` required for all objects in a hierarchy or only for the ones that should be ordered? If it is the latter, this can stay just as `json` since only each item needs to be ordered.

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp [hunk 4]
```diff
@@ -104,16 +104,17 @@ class GeometryHierarchyMapJsonConverter {
 // implementations
 
 template <typename value_t>
-nlohmann::json GeometryHierarchyMapJsonConverter<value_t>::toJson(
+nlohmann::ordered_json GeometryHierarchyMapJsonConverter<value_t>::toJson(
```
Comment (msmk0) [inline Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:107]: See above.

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp [hunk 5]
```diff
@@ -104,16 +104,17 @@ class GeometryHierarchyMapJsonConverter {
...
-  nlohmann::json encoded = nlohmann::json::object();
+  nlohmann::ordered_json encoded = nlohmann::json::object();
+  ;
   encoded[kHeaderKey] = {
       {"format-version", kFormatVersion},
       {"value-identifier", m_valueIdentifier},
   };
   // encode entries
-  nlohmann::json entries = nlohmann::json::array();
+  nlohmann::ordered_json entries = nlohmann::json::array();
```
Comment (msmk0) [inline Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:117]: See above, but a JSON array is always ordered anyways.

**Review by msmk0 — CHANGES_REQUESTED** [review]
Looks Ok. Could you also update `CMakeLists.txt` and `thirdparty/README.md` to reflect the increased version requirement for the JSON library.

**Comment by msmk0** [comment]
> Once it is in I will update the other PR to take those change into account

@Corentin-Allaire My apologies, I did not see that @asalzburger had already reviewed this.

File: thirdparty/nlohmann_json/cmake/download_test_data.cmake (hunk 2 above)
Comment (Corentin-Allaire) [reply thirdparty/nlohmann_json/cmake/download_test_data.cmake:1]: I just copied the directory whithout thinking much about it but you are right I should remove probably this. Let me check

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp (hunk 1 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:111]: Yes I will remove it, thanks.

**Comment by Corentin-Allaire** [comment]
> > Once it is in I will update the other PR to take those change into account
> 
> @Corentin-Allaire My apologies, I did not see that @asalzburger had already reviewed this.

Your comment are much appreciated. I will have a look at them

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp (hunk 3 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:55]: I think if you do `json = ordered_json` you lose the ordering, so the final object you return has to be of type `ordered_json`. To be honest I haven't really found documentation on `ordered_json` but I just tested it and changing back to `json` doesn't not work in our case.

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp (hunk 4 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:107]: See above

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp (hunk 5 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:117]: In that case if you use `nlohmann::json` you cannot use `.push_back` anymore. `json.push_back(ordered_json)` does not compile.

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp (hunk 3 above)
Comment (msmk0) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:55]: Ok, thanks for checking.

**Review by msmk0 — APPROVED** [review]
Thanks for answering my comments. I will merge this.
